GPS updates frozen actors that change owner

RoosterDragon opened the pull request:
There are two main changes in this PR:
- A new trait, `FrozenUnderFogUpdatedByGps`, replaces `GpsRemoveFrozenActor`. It supports removing frozen actors that are sold or die but also now supports refreshing the frozen actor when ownership changes. This brings it in line with the GPS dot which changes colour as required.
- Remove the `IRemoveFrozenActor` interface. This caused `FrozenActorLayer.Tick` to be quite slow. The new trait only needs to run when the GPS state changes, rather than every tick. This is faster.

I would test the following:
- When a GPS is launched, that frozen actors are updated as follows:
  - A frozen actor for a sold building is removed.
  - A frozen actor for a destroyed building is removed.
  - A frozen actor for a building that was captured is updated with new colours.
- With an active GPS, check that frozen actors are updated live for the above three scenarios.

---

On the RA shellmap, reduces CPU time spent in `FrozenActorLayer.Tick` from 1.39% to 0.83%.

Suggested changelog: GPS updates frozen actors that change owner

RoosterDragon commented on OpenRA.Mods.RA/Traits/GpsRemoveFrozenActor.cs:
```diff
@@ -35,21 +34,31 @@ public GpsRemoveFrozenActor(Actor self, GpsRemoveFrozenActorInfo info)
 					continue;
 
-				if (w.Launched)
-					return true;
```
Attention: I have changed this check from `Launched` to `Granted` which I believe is more correct and matches the description of the trait (an *active* GPS power).

pchote commented on OpenRA.Mods.RA/Traits/GpsRemoveFrozenActor.cs:
```diff
@@ -35,21 +34,31 @@ public GpsRemoveFrozenActor(Actor self, GpsRemoveFrozenActorInfo info)
 		}
 
-		public bool RemoveActor(Actor self, Player owner)
+		public void Tick(Actor self)
```
You can't guarantee this to be called after an actor is killed.  You should use INotifyRemovedFromWorld or similar instead.

RoosterDragon replied:
I am already implementing `INotifyKilled` which seems to pick up sold and destroyed buildings just fine.

pchote replied:
Should be able to remove this one then.

RoosterDragon replied:
We need this one to erase any frozen actors that are already dead when the GPS first activates.

RoosterDragon replied:
Nope I'm being dumb here. I'll probably need to wire up a way for the GPS to signal when the GPS grant status changes so this trait can re-evaluate stuff.

pchote replied:
`INotifyGPSStateChanged`?

RoosterDragon committed:
Implement equality on TraitPair.

RoosterDragon committed:
Create FrozenActor.RefreshState method for better encapsulation and reuse of this logic.

RoosterDragon committed:
Replace IRemoveFrozenActor with FrozenActorLayer.Remove.

The IRemoveFrozenActor interface is replaced with a Remove method on FrozenActorLayer. IRemoveFrozenActor is a performance problem for FrozenActorLayer.Tick as it incurs a large cache miss penalty in order to load and enumerate the array of these interfaces for every frozen actor. Instead, we invert control and allow traits to remove frozen actors directly which eliminates the performance penalty.

RoosterDragon committed:
Introduce FrozenUnderFogUpdatedByGps and IOnGpsRefreshed.

This trait uses the new interface to register with the GpsWatcher to get notified when the GPS state changes. It needs to register as if the actor is destroyed, the GpsWatcher can no longer query actors in the world to get the trait. This is no good as the trait needs to perform actions after the death of its parent actor.

When the GPS is enabled, the trait will track ownership changes and update the frozen actor (just like if there was only a GPS dot, it would change colour) and it will also remove frozen actors if the real actor dies or is sold.

RoosterDragon committed:
Add an upgrade rule for FrozenUnderFogUpdatedByGps.

RoosterDragon commented:
This has grown in complexity somewhat. The trait needing to do things after the actor dies makes it more difficult than usual to notify it about things.

On the plus side, this change means I can support the GPS updating a frozen actor that changes ownership, which is nice.

I have updated the opening post.

Mailaender commented:
Needs a rebase now.

RoosterDragon commented:
Rebased.

abcdefg30 commented:
Needs a rebase again. :/

RoosterDragon commented:
Rebased.

Mailaender commented:
> A frozen actor for a destroyed building is removed.

Works and I can't spot any related regressions. :white_check_mark:

obrakmann commented:
Looks good to me, :+1: 

Please rebase and update the engine version in the upgrade rule so we can get this merged.

RoosterDragon commented:
Rebased and date amended to today.

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/ee752baeb16de71a29ed395c617c0f9272795c68)
